apps/ledger-live-mobile/src/components/Web3AppWebview/useSwapHeaderNavigation.tsx

```diff
@@ -0,0 +1,119 @@
+      return t("transfer.swap2.twoStepApproval.title");
+    case SwapWebviewAllowedPageNames.QuotesList:
+      return t("transfer.swap2.quotesList.title");
+    default:
+      return t("transfer.swap2.form.title");
+  }
+}
+
+export function useSwapHeaderNavigation(webviewRef: React.RefObject<WebView>) {
+  const navigation = useNavigation();
+  const { t } = useTranslation();
+  const track = useTrack();
+
+  const { isSwapTabScreen, swapTabScreen } = useIsSwapTab();
+
+  const navigateToSwapHistory = useCallback(() => {
+    track("button_clicked", {
+      button: "SwapHistory",
+      page: ScreenName.SwapTab,
+      swapVersion: SWAP_VERSION,
+    });
+
+    navigation.navigate(ScreenName.SwapHistory);
+  }, [navigation, track]);
+
+  const goBackWebView = useCallback(
+    (currentWebviewPage?: SwapWebviewAllowedPageNames) => {
+      track("button_clicked", {
+        button: "SwapWebviewBack",
+        page: ScreenName.SwapTab,
+        webviewPage: currentWebviewPage,
+        swapVersion: SWAP_VERSION,
+      });
+
+      webviewRef.current?.goBack();
+    },
+    [webviewRef, track],
+  );
+
+  const goBackNative = useCallback(() => {
+    track("button_clicked", {
+      button: "SwapNativeBack",
+      page: ScreenName.SwapTab,
+      swapVersion: SWAP_VERSION,
+    });
+
+    navigation.goBack();
+  }, [navigation, track]);
+
+  if (isSwapTabScreen) {
+    const webviewParams = swapTabScreen?.params;
+
+    const webviewCanGoBack = (webviewParams as DetailsSwapParamList | DefaultAccountSwapParamList)
+      ?.swapNavigationParams?.canGoBack;
+
+    const webviewCurrentPage = (webviewParams as DetailsSwapParamList | DefaultAccountSwapParamList)
+      ?.swapNavigationParams?.page;
+
+    if (webviewCanGoBack && webviewCurrentPage !== SwapWebviewAllowedPageNames.AccountSelection) {
+      navigation.setOptions({
+        headerLeft: () => (
+          <Flex p={6}>
+            <Touchable
+              touchableTestID="NavigationHeaderClose"
+              onPress={() => goBackWebView(webviewCurrentPage)}
+            >
+              <Icons.ArrowLeft color={"neutral.c100"} />
+            </Touchable>
+          </Flex>
+        ),
+        headerRight: undefined,
+        headerTitle: getScreenTitle({ t, webviewCurrentPage }),
+      });
+    } else {
+      navigation.setOptions({
+        headerLeft: () => (
+          <Flex p={6}>
+            <Touchable touchableTestID="NavigationHeaderClose" onPress={goBackNative}>
+              <Icons.ArrowLeft color={"neutral.c100"} />
+            </Touchable>
+          </Flex>
+        ),
+        headerRight: () => (
+          <Flex p={6}>
+            <Touchable
+              touchableTestID="NavigationHeaderSwapHistory"
+              onPress={navigateToSwapHistory}
+            >
+              <Icons.Clock color={"neutral.c100"} />
+            </Touchable>
+          </Flex>
+        ),
+        headerTitle: t("transfer.swap2.form.title"),
+      });
+    }
+  }
```

Comment:
I'm not sure we're supposed to call setOptions in the render pass, we should probably move all of it in a useEffect, wdyt ?